File: xim/poc/ActorModel.kt
package xim.poc

import xim.math.Vector3f
import xim.poc.camera.CameraReference
import xim.poc.gl.MeshBuffer
import xim.resource.*

enum class FootState {
    Grounded,
    Lifting,
    Ungrounded,
    Landing,
    ;

    fun next(isTouching: Boolean): FootState {
        return when (this) {
            Grounded -> if (isTouching) { Grounded } else { Lifting }
            Lifting -> if (isTouching) { Grounded } else { Ungrounded }
            Landing -> if (isTouching) { Grounded } else { Ungrounded }
            Ungrounded -> if (isTouching) { Landing } else { Ungrounded }
        }
    }
}

class CustomModelSettings(
    var blurConfig: BlurConfig? = null,
    var scale: Float = 1f,
    var effectScale: Float? = null,
    var hideMain: Boolean = false,
)

data class SlotVisibilityOverride(val slot: Int, val hidden: Boolean, val ifEngaged: Boolean)

class ModelSlotVisibilityState {

    private val visibilityState = HashMap<Int, SlotVisibilityOverride>()

    fun apply(slotVisibilityOverride: SlotVisibilityOverride) {
        visibilityState[slotVisibilityOverride.slot] = slotVisibilityOverride
    }

    fun getOverrides(): List<SlotVisibilityOverride> {
        return visibilityState.values.toList()
    }

}

class ModelLock(var duration: Float, val validityCheck: () -> Boolean = { true })

class ActorModel(val actor: Actor, val model: Model) {

    private var skeletonInstance: SkeletonInstance? = null
    val skeletonAnimationCoordinator = SkeletonAnimationCoordinator()
    val skeletonAnimationThrottler = SkeletonAnimationThrottler(actor)

    private val defaultVisibilityStates = ModelSlotVisibilityState()
    private val hiddenModelSlotIds = HashSet<Int>()

    private var leftFootState: FootState = FootState.Grounded
    private var rightFootState: FootState = FootState.Grounded

    private var animationLocks = ArrayList<ModelLock>()
    private var movementLocks = ArrayList<ModelLock>()
    private var facingLocks = ArrayList<ModelLock>()

    private var displayRangedDuration = 0f

    var customModelSettings = CustomModelSettings()

    var idleAnimationMode = 0
    var battleAnimationMode = 0
    var walkingAnimationMode = 0
    var runningAnimationMode = 0

    fun update(elapsedFrames: Float) {
        if (skeletonInstance == null) {
            val resource = model.getSkeletonResource() ?: return
            skeletonInstance = SkeletonInstance(resource)
        }

        skeletonAnimationCoordinator.update(elapsedFrames)
        updateLocks(elapsedFrames, movementLocks)
        updateLocks(elapsedFrames, animationLocks)
        updateLocks(elapsedFrames, facingLocks)
        displayRangedDuration -= elapsedFrames

        leftFootState = leftFootState.next(skeletonInstance!!.isLeftFootTouchingGround())
        rightFootState = rightFootState.next(skeletonInstance!!.isRightFootTouchingGround())
    }

    private fun updateLocks(elapsedFrames: Float, locks: ArrayList<ModelLock>) {
        locks.forEach { it.duration -= elapsedFrames }
        locks.removeAll { it.duration <= 0f || !it.validityCheck.invoke() }
    }

    fun getMeshes(): List<MeshBuffer> {
        val directories = model.getMeshResources()
        val resources = directories.flatMap { it.collectByTypeRecursive(SkeletonMeshResource::class) }.toMutableList()

        val hiddenIds = getHiddenSlotIds().map { DatId("wep$it") }
        resources.removeAll { hiddenIds.contains(it.id) }

        val occluded = resources.map { it.occlusionType }.toSet()
        return resources.flatMap { it.meshes }.filter { !isOccluded(it, occluded) }
    }

    fun getHiddenModelSlots(): Set<ItemModelSlot> {
        return getHiddenSlotIds().mapNotNull { when(it) {
            0 -> ItemModelSlot.Main
            1 -> ItemModelSlot.Sub
            2 -> ItemModelSlot.Range
            else -> null
        } }.toSet()
    }

    fun lockAnimation(duration: Float) {
        lockAnimation(ModelLock(duration))
    }

    fun lockAnimation(modelLock: ModelLock) {
        animationLocks += modelLock
    }

    fun isAnimationLocked(): Boolean {
        return animationLocks.isNotEmpty()
    }

    fun lockMovement(duration: Float) {
        lockMovement(ModelLock(duration))
    }

    fun lockMovement(modelLock: ModelLock) {
        movementLocks += modelLock
    }

    fun isMovementLocked() : Boolean {
        return movementLocks.isNotEmpty()
    }

    fun lockFacing(duration: Float) {
        lockFacing(ModelLock(duration))
    }

    fun lockFacing(modelLock: ModelLock) {
        facingLocks += modelLock
    }

    fun isFacingLocked() : Boolean {
        return facingLocks.isNotEmpty()
    }

    fun clearAnimations() {
        skeletonAnimationCoordinator.clear()
    }

    fun getSkeleton(): SkeletonInstance? {
        return skeletonInstance
    }

    fun setSkeletonAnimation(datId: DatId, animationDirs: List<DirectoryResource>, loopParams: LoopParams = LoopParams.lowPriorityLoop(), transitionParams: TransitionParams, modelSlotVisibilityState: ModelSlotVisibilityState? = null) {
        val resources = fetchAnimations(datId, animationDirs)

        if (transitionParams.inBetween != null) {
            transitionParams.resolvedInBetween = fetchAnimations(transitionParams.inBetween, animationDirs).associateBy { it.id.finalDigit() ?: 0 }
        }

        skeletonAnimationCoordinator.registerAnimation(resources, loopParams, transitionParams, modelSlotVisibilityState)
    }

    fun transitionToMoving(datId: DatId, animationDirs: List<DirectoryResource>, transitionParams: TransitionParams) {
        setSkeletonAnimation(datId, animationDirs, LoopParams.lowPriorityLoop(), transitionParams)
    }

    fun transitionToIdleOnStopMoving(datId: DatId, animationDirs: List<DirectoryResource>) {
        val idleResources = fetchAnimations(datId, animationDirs)
        skeletonAnimationCoordinator.registerIdleAnimation(idleResources, requireTransitionOut = false)
    }

    fun transitionToIdleOnCompleted(idleId: DatId, animationDirs: List<DirectoryResource>) {
        val alreadyIdle = skeletonAnimationCoordinator.animations.filterNotNull()
            .mapNotNull { it.currentAnimation?.animation?.id }
            .all { it.parameterizedMatch(idleId) }

        if (alreadyIdle) { return }

        if (!skeletonAnimationCoordinator.hasCompleteTransitionOutAnimations()) { return }

        val idleResources = fetchAnimations(idleId, animationDirs)
        skeletonAnimationCoordinator.registerIdleAnimation(idleResources, requireTransitionOut = true)
    }

    fun forceTransitionToIdle(idleId: DatId, transitionTime: Float, animationDirs: List<DirectoryResource>) {
        displayRangedDuration = 0f
        val idleResources = fetchAnimations(idleId, animationDirs)
        skeletonAnimationCoordinator.registerAnimation(idleResources,
            LoopParams.lowPriorityLoop(),
            TransitionParams(transitionInTime = transitionTime, transitionOutTime = 0f)
        )
    }

    private fun fetchAnimations(datId: DatId, animationDirs: List<DirectoryResource>): List<SkeletonAnimationResource> {
        return animationDirs.map { it.root() }
            .flatMap { it.collectByTypeRecursive(SkeletonAnimationResource::class) }
            .filter { it.id.parameterizedMatch(datId) }
            .distinctBy { it.id }
    }

    fun getJointPosition(standardPosition: Int): Vector3f? {
        val skeleton = getSkeleton() ?: return null
        return skeleton.getStandardJointPosition(standardPosition)
    }

    fun getFootInfoDefinition() : InfoDefinition? {
        return model.getMovementInfo()
    }

    fun getCollidingFoot(): StandardPosition? {
        return if (leftFootState == FootState.Landing) {
            StandardPosition.LeftFoot
        } else if (rightFootState == FootState.Landing) {
            StandardPosition.RightFoot
        } else {
            null
        }
    }

    fun displayRanged(duration: Float) {
        if (hiddenModelSlotIds.contains(2)) { return }
        displayRangedDuration = duration
    }

    fun getBlurConfig(): BlurConfig? {
        return customModelSettings.blurConfig ?: model.getBlurConfig()
    }

    fun setDefaultModelVisibility(slotVisibilityOverride: SlotVisibilityOverride) {
        defaultVisibilityStates.apply(slotVisibilityOverride)
    }

    fun toggleModelVisibility(slot: Int, hidden: Boolean) {
        if (hidden) { hiddenModelSlotIds += slot } else { hiddenModelSlotIds -= slot }
        if (slot == 2) { displayRangedDuration = 0f }
    }

    private fun isOccluded(meshBuffer: MeshBuffer, occlusion: Set<Int>): Boolean {
        val renderProperties = meshBuffer.skeletalMeshProperties ?: return false

        // TODO - do occlusions 0x01, 0x11, 0x21, and 0x31 do anything?
        return when (renderProperties.displayTypeFlag) {
            0 -> false
            /* hair 1*/ 1 -> occlusion.contains(0x02) || occlusion.contains(0x03) || occlusion.contains(0x04) || occlusion.contains(0x05) || occlusion.contains(0x06)
            /* hair 2*/ 2 -> occlusion.contains(0x04) || occlusion.contains(0x05) || occlusion.contains(0x06)
            /* hair 3*/ 3 -> occlusion.contains(0x04) || occlusion.contains(0x05) || occlusion.contains(0x06)
            /* face  */ 4 -> occlusion.contains(0x05)
            /* wrist */ 5 -> occlusion.contains(0x12)
            /* pants */ 6 -> occlusion.contains(0x32)
            /* shins */ 7 -> occlusion.contains(0x22)
            else -> throw IllegalStateException("Unknown display type: ${renderProperties.displayTypeFlag}")
        }

    }

    private fun getHiddenSlotIds(): Set<Int> {
        if (actor.getMount() != null) { return setOf(0, 1, 2) }

        // Precedence is: ActorModel visibility > Display-Ranged > Effect Visibility.
        // Ranged is hidden by default.
        val hiddenSlots = mutableSetOf(2)

        val engaged = actor.isDisplayEngagedOrEngaging()

        val overrideStates = defaultVisibilityStates.getOverrides() +
                skeletonAnimationCoordinator.getSlotVisibilityOverrides().flatMap { it.getOverrides() }

        for (override in overrideStates)  {
            if (override.ifEngaged && !engaged) { continue }
            if (override.hidden) {
                hiddenSlots += override.slot
            } else {
                hiddenSlots -= override.slot
            }
        }

        hiddenSlots.addAll(hiddenModelSlotIds)

        if (!hiddenSlots.contains(2) && displayRangedDuration > 0f) { return setOf(0, 1) }

        return hiddenSlots
    }

    fun getMainAttackIds(appearanceState: Int): List<DatId> {
        val dir = model.getMainBattleAnimationDirectory() ?: return emptyList()

        if (appearanceState != 0) {
            val ids = collectIdsByPrefix(dir, "a${appearanceState}i")
            if (ids.isNotEmpty()) { return ids }
        }

        return collectIdsByPrefix(dir, "ati")
    }

    fun getSubAttackIds(appearanceState: Int): List<DatId> {
        val dir = model.getSubBattleAnimationDirectory() ?: return emptyList()

        if (appearanceState != 0) {
            val ids = collectIdsByPrefix(dir, "b${appearanceState}i")
            if (ids.isNotEmpty()) { return ids }
        }

        return collectIdsByPrefix(dir, "bti")
    }

    private fun collectIdsByPrefix(dir: DirectoryResource, prefix: String): List<DatId> {
        return dir.collectByTypeRecursive(EffectRoutineResource::class)
            .map { it.id }
            .filter { it.id.startsWith(prefix) }
    }

}

File: lib/runtime/actorModel.ts
import {
  DatId,
  type DirectoryResource,
  type InfoDefinition,
  SkeletonAnimationResource,
  SkeletonMeshResource,
} from '~/lib/resource/datResource'
import { SkeletonInstance } from '~/lib/resource/skeletonInstance'
import { datIdParameterizedMatch } from '~/lib/runtime/datId'
import type { BlurConfig, Model } from '~/lib/runtime/model'
import {
  LoopParams,
  type TransitionParams,
  SkeletonAnimationCoordinator,
} from '~/lib/runtime/skeletonAnimator'
import { collectByTypeRecursive } from '~/lib/runtime/resourceTree'

const identityRotation = { x: 0, y: 0, z: 0, w: 1 } as const

export enum FootState {
  Grounded = 'Grounded',
  Lifting = 'Lifting',
  Ungrounded = 'Ungrounded',
  Landing = 'Landing',
}

function nextFootState(state: FootState, isTouching: boolean): FootState {
  if (state === FootState.Grounded) {
    return isTouching ? FootState.Grounded : FootState.Lifting
  }
  if (state === FootState.Lifting || state === FootState.Landing) {
    return isTouching ? FootState.Grounded : FootState.Ungrounded
  }
  return isTouching ? FootState.Landing : FootState.Ungrounded
}

export class CustomModelSettings {
  blurConfig: BlurConfig | null = null
  scale = 1
  effectScale: number | null = null
  hideMain = false
}

export class SlotVisibilityOverride {
  readonly slot: number
  readonly hidden: boolean
  readonly ifEngaged: boolean

  constructor(slot: number, hidden: boolean, ifEngaged: boolean) {
    this.slot = slot
    this.hidden = hidden
    this.ifEngaged = ifEngaged
  }
}

export class ModelSlotVisibilityState {
  private readonly visibilityState = new Map<number, SlotVisibilityOverride>()

  apply(slotVisibilityOverride: SlotVisibilityOverride): void {
    this.visibilityState.set(slotVisibilityOverride.slot, slotVisibilityOverride)
  }

  getOverrides(): SlotVisibilityOverride[] {
    return Array.from(this.visibilityState.values())
  }
}

export class ModelLock {
  duration: number
  readonly validityCheck: () => boolean

  constructor(duration: number, validityCheck: () => boolean = () => true) {
    this.duration = duration
    this.validityCheck = validityCheck
  }
}

export interface RuntimeActor {
  isDisplayEngagedOrEngaging(): boolean
  getMount(): unknown | null
}

export class ActorModel {
  readonly actor: RuntimeActor
  readonly model: Model

  private skeletonInstance: SkeletonInstance | null = null
  readonly skeletonAnimationCoordinator = new SkeletonAnimationCoordinator()

  private readonly defaultVisibilityStates = new ModelSlotVisibilityState()
  private readonly hiddenModelSlotIds = new Set<number>()

  private leftFootState: FootState = FootState.Grounded
  private rightFootState: FootState = FootState.Grounded

  private readonly animationLocks: ModelLock[] = []
  private readonly movementLocks: ModelLock[] = []
  private readonly facingLocks: ModelLock[] = []

  private displayRangedDuration = 0

  customModelSettings = new CustomModelSettings()
  idleAnimationMode = 0
  battleAnimationMode = 0
  walkingAnimationMode = 0
  runningAnimationMode = 0

  constructor(actor: RuntimeActor, model: Model) {
    this.actor = actor
    this.model = model
  }

  update(elapsedFrames: number): void {
    if (!this.skeletonInstance) {
      const resource = this.model.getSkeletonResource()
      if (!resource) {
        return
      }
      this.skeletonInstance = new SkeletonInstance(resource)
      this.skeletonInstance.tPose()
    }

    this.skeletonAnimationCoordinator.update(elapsedFrames)
    this.updateLocks(elapsedFrames, this.movementLocks)
    this.updateLocks(elapsedFrames, this.animationLocks)
    this.updateLocks(elapsedFrames, this.facingLocks)

    this.displayRangedDuration -= elapsedFrames
    this.updateSkeletonPose()

    this.leftFootState = nextFootState(this.leftFootState, this.skeletonInstanceIsLeftFootTouchingGround())
    this.rightFootState = nextFootState(this.rightFootState, this.skeletonInstanceIsRightFootTouchingGround())
  }

  getSkeleton(): SkeletonInstance | null {
    return this.skeletonInstance
  }

  getMeshResources(): SkeletonMeshResource[] {
    const directories = this.model.getMeshResources()
    const resources = directories.flatMap((directory) => collectByTypeRecursive(directory, SkeletonMeshResource))
    const hiddenIds = new Set(Array.from(this.getHiddenSlotIds()).map((slot) => `wep${slot}`))
    return resources.filter((resource) => !hiddenIds.has(resource.id.id))
  }

  getHiddenSlotIds(): Set<number> {
    if (this.actor.getMount() !== null) {
      return new Set([0, 1, 2])
    }

    const hiddenSlots = new Set<number>([2])
    const engaged = this.actor.isDisplayEngagedOrEngaging()

    const overrideStates = [
      ...this.defaultVisibilityStates.getOverrides(),
      ...this.skeletonAnimationCoordinator
        .getSlotVisibilityOverrides()
        .flatMap((state) => state.getOverrides()),
    ]

    for (const slotOverride of overrideStates) {
      if (slotOverride.ifEngaged && !engaged) {
        continue
      }

      if (slotOverride.hidden) {
        hiddenSlots.add(slotOverride.slot)
      } else {
        hiddenSlots.delete(slotOverride.slot)
      }
    }

    for (const hiddenSlot of this.hiddenModelSlotIds) {
      hiddenSlots.add(hiddenSlot)
    }

    if (!hiddenSlots.has(2) && this.displayRangedDuration > 0) {
      return new Set([0, 1])
    }

    return hiddenSlots
  }

  setSkeletonAnimation(
    datId: DatId,
    animationDirs: readonly DirectoryResource[],
    loopParams: LoopParams = LoopParams.lowPriorityLoop(),
    transitionParams: TransitionParams,
    modelSlotVisibilityState: ModelSlotVisibilityState | null = null,
  ): void {
    const resources = this.fetchAnimations(datId, animationDirs)

    if (transitionParams.inBetween) {
      const inBetweenResources = this.fetchAnimations(transitionParams.inBetween, animationDirs)
      transitionParams.resolvedInBetween = new Map(
        inBetweenResources.map((resource) => {
          const key = Number.parseInt(resource.id.id.at(-1) ?? '0', 36)
          return [Number.isNaN(key) ? 0 : key, resource] as const
        }),
      )
    }

    this.skeletonAnimationCoordinator.registerAnimation(resources, loopParams, transitionParams, modelSlotVisibilityState)
  }

  transitionToMoving(datId: DatId, animationDirs: readonly DirectoryResource[], transitionParams: TransitionParams): void {
    this.setSkeletonAnimation(datId, animationDirs, LoopParams.lowPriorityLoop(), transitionParams)
  }

  transitionToIdleOnStopMoving(datId: DatId, animationDirs: readonly DirectoryResource[]): void {
    const idleResources = this.fetchAnimations(datId, animationDirs)
    this.skeletonAnimationCoordinator.registerIdleAnimation(idleResources, false)
  }

  transitionToIdleOnCompleted(datId: DatId, animationDirs: readonly DirectoryResource[]): void {
    const alreadyIdle = this.skeletonAnimationCoordinator.animations
      .filter((animator): animator is NonNullable<(typeof this.skeletonAnimationCoordinator.animations)[number]> => animator !== null)
      .map((animator) => animator.currentAnimation?.animation.id)
      .every((currentId) => (currentId ? datIdParameterizedMatch(currentId, datId) : false))

    if (alreadyIdle || !this.skeletonAnimationCoordinator.hasCompleteTransitionOutAnimations()) {
      return
    }

    const idleResources = this.fetchAnimations(datId, animationDirs)
    this.skeletonAnimationCoordinator.registerIdleAnimation(idleResources, true)
  }

  forceTransitionToIdle(idleId: DatId, transitionTime: number, animationDirs: readonly DirectoryResource[]): void {
    this.displayRangedDuration = 0
    const idleResources = this.fetchAnimations(idleId, animationDirs)
    this.skeletonAnimationCoordinator.registerAnimation(
      idleResources,
      LoopParams.lowPriorityLoop(),
      { transitionInTime: transitionTime, transitionOutTime: 0, inBetween: null, eagerTransitionOut: false, resolvedInBetween: null },
      null,
    )
  }

  getJointPosition(standardPosition: number): { x: number, y: number, z: number } | null {
    const skeleton = this.getSkeleton()
    if (!skeleton) {
      return null
    }

    try {
      return skeleton.getStandardJointPosition(standardPosition)
    } catch {
      return null
    }
  }

  getFootInfoDefinition(): InfoDefinition | null {
    return this.model.getMovementInfo()
  }

  displayRanged(duration: number): void {
    if (this.hiddenModelSlotIds.has(2)) {
      return
    }
    this.displayRangedDuration = duration
  }

  getBlurConfig(): BlurConfig | null {
    return this.customModelSettings.blurConfig ?? this.model.getBlurConfig()
  }

  setDefaultModelVisibility(slotVisibilityOverride: SlotVisibilityOverride): void {
    this.defaultVisibilityStates.apply(slotVisibilityOverride)
  }

  toggleModelVisibility(slot: number, hidden: boolean): void {
    if (hidden) {
      this.hiddenModelSlotIds.add(slot)
    } else {
      this.hiddenModelSlotIds.delete(slot)
    }

    if (slot === 2) {
      this.displayRangedDuration = 0
    }
  }

  lockAnimation(duration: number | ModelLock): void {
    this.animationLocks.push(duration instanceof ModelLock ? duration : new ModelLock(duration))
  }

  isAnimationLocked(): boolean {
    return this.animationLocks.length > 0
  }

  lockMovement(duration: number | ModelLock): void {
    this.movementLocks.push(duration instanceof ModelLock ? duration : new ModelLock(duration))
  }

  isMovementLocked(): boolean {
    return this.movementLocks.length > 0
  }

  lockFacing(duration: number | ModelLock): void {
    this.facingLocks.push(duration instanceof ModelLock ? duration : new ModelLock(duration))
  }

  isFacingLocked(): boolean {
    return this.facingLocks.length > 0
  }

  clearAnimations(): void {
    this.skeletonAnimationCoordinator.clear()
  }

  private updateLocks(elapsedFrames: number, locks: ModelLock[]): void {
    for (const lock of locks) {
      lock.duration -= elapsedFrames
    }

    for (let i = locks.length - 1; i >= 0; i -= 1) {
      const lock = locks[i]
      if (lock && (lock.duration <= 0 || !lock.validityCheck())) {
        locks.splice(i, 1)
      }
    }
  }

  private fetchAnimations(datId: DatId, animationDirs: readonly DirectoryResource[]): SkeletonAnimationResource[] {
    const allAnimations = animationDirs.flatMap((root) => collectByTypeRecursive(root, SkeletonAnimationResource))
    const matched = allAnimations.filter((resource) => datIdParameterizedMatch(resource.id, datId))
    const byId = new Map<string, SkeletonAnimationResource>()

    for (const resource of matched) {
      byId.set(resource.id.id, resource)
    }

    return Array.from(byId.values())
  }

  private updateSkeletonPose(): void {
    const skeleton = this.skeletonInstance
    if (!skeleton) {
      return
    }

    const resourceJoints = skeleton.resource.joints

    for (let i = 0; i < skeleton.joints.length; i += 1) {
      const joint = skeleton.joints[i]
      const jointDef = resourceJoints[i]
      if (!joint || !jointDef) {
        continue
      }

      const animationTransform = this.skeletonAnimationCoordinator.getJointTransform(i)
      const localRotation = normalizeQuaternion(
        multiplyQuaternion(animationTransform?.rotation ?? identityRotation, jointDef.rotation),
      )
      const localTranslation = {
        x: finiteNumber(jointDef.translation.x) + finiteNumber(animationTransform?.translation.x),
        y: finiteNumber(jointDef.translation.y) + finiteNumber(animationTransform?.translation.y),
        z: finiteNumber(jointDef.translation.z) + finiteNumber(animationTransform?.translation.z),
      }

      if (joint.parentIndex < 0) {
        joint.worldPosition.x = localTranslation.x
        joint.worldPosition.y = localTranslation.y
        joint.worldPosition.z = localTranslation.z
        joint.worldRotation.x = localRotation.x
        joint.worldRotation.y = localRotation.y
        joint.worldRotation.z = localRotation.z
        joint.worldRotation.w = localRotation.w
        continue
      }

      const parent = skeleton.joints[joint.parentIndex]
      if (!parent) {
        continue
      }

      const rotatedTranslation = rotateVector(parent.worldRotation, localTranslation)
      joint.worldPosition.x = parent.worldPosition.x + rotatedTranslation.x
      joint.worldPosition.y = parent.worldPosition.y + rotatedTranslation.y
      joint.worldPosition.z = parent.worldPosition.z + rotatedTranslation.z

      const worldRotation = normalizeQuaternion(multiplyQuaternion(parent.worldRotation, localRotation))
      joint.worldRotation.x = worldRotation.x
      joint.worldRotation.y = worldRotation.y
      joint.worldRotation.z = worldRotation.z
      joint.worldRotation.w = worldRotation.w
    }
  }

  private skeletonInstanceIsLeftFootTouchingGround(): boolean {
    return this.leftFootState === FootState.Grounded || this.leftFootState === FootState.Landing
  }

  private skeletonInstanceIsRightFootTouchingGround(): boolean {
    return this.rightFootState === FootState.Grounded || this.rightFootState === FootState.Landing
  }
}

function multiplyQuaternion(
  a: { x: number, y: number, z: number, w: number },
  b: { x: number, y: number, z: number, w: number },
): { x: number, y: number, z: number, w: number } {
  return {
    w: a.w * b.w - a.x * b.x - a.y * b.y - a.z * b.z,
    x: a.w * b.x + a.x * b.w + a.y * b.z - a.z * b.y,
    y: a.w * b.y - a.x * b.z + a.y * b.w + a.z * b.x,
    z: a.w * b.z + a.x * b.y - a.y * b.x + a.z * b.w,
  }
}

function conjugateQuaternion(q: { x: number, y: number, z: number, w: number }): { x: number, y: number, z: number, w: number } {
  return {
    x: -q.x,
    y: -q.y,
    z: -q.z,
    w: q.w,
  }
}

function normalizeQuaternion(q: { x: number, y: number, z: number, w: number }): { x: number, y: number, z: number, w: number } {
  const magnitude = Math.hypot(q.x, q.y, q.z, q.w)
  if (!Number.isFinite(magnitude) || magnitude <= 1e-7) {
    return { ...identityRotation }
  }

  return {
    x: q.x / magnitude,
    y: q.y / magnitude,
    z: q.z / magnitude,
    w: q.w / magnitude,
  }
}

function finiteNumber(value: number | undefined): number {
  if (value === undefined || !Number.isFinite(value)) {
    return 0
  }
  return value
}

function rotateVector(
  rotation: { x: number, y: number, z: number, w: number },
  vector: { x: number, y: number, z: number },
): { x: number, y: number, z: number } {
  const q = normalizeQuaternion(rotation)
  const vecQuat = { x: vector.x, y: vector.y, z: vector.z, w: 0 }
  const rotated = multiplyQuaternion(multiplyQuaternion(q, vecQuat), conjugateQuaternion(q))
  return {
    x: rotated.x,
    y: rotated.y,
    z: rotated.z,
  }
}
